lua/neogit/popups/branch/actions.lua

```diff
@@ -22,6 +22,40 @@ local function parse_remote_branch_name(ref)
   return remote, branch_name
 end
 
+M.spin_off_branch = operation("spin_off_branch", function()
+  if #git.repo.staged.items > 0 or #git.repo.unstaged.items > 0 then
+    notif.create("Staying on current branch as there are uncommitted changes.", vim.log.levels.INFO)
+    return
+  end
+
+  local name = git.branch.create()
+
+  local upstream = git.repo.upstream.ref
+  if upstream then
+    git.cli.reset.hard.args(upstream).call_sync()
+  end
+
+  git.cli.checkout.branch(name).call_sync()
```

Comment:
Before resetting the existing branch and checking out the new one, Magit adds a reflog entry for the action:

https://github.com/magit/magit/blob/main/lisp/magit-branch.el#L517
```elisp
(magit-call-git "update-ref" "-m" (format "reset: moving to %s" base) (concat "refs/heads/" current) base)
```

Which, roughly translates to lua as:
```lua
local base = "main"
local current = "feature-branch"

magit_call_git("git", "update-ref", "-m", string.format("reset: moving to %s", base), "refs/heads/" .. current, base)
```

Reply:
update: I added `git.log.update_ref(from, to)` for this

Reply:
Another thing I misunderstood 😅

In Magit, it seems that the ref log is updated _instead_ of hard resetting when the new branch should be checked out, and is to ensure uncommitted changes aren't lost. This would be the default behaviour for `spin_off`, and when there are uncommitted changes during `spin_out`.

I'll update to align with the above.

Reply:
Ahh, I didn't realize it was instead of, but it looks like you're right. Nice work